lib/SYCL/SYCLArgsFlattening.cpp

```diff
@@ -332,12 +334,48 @@ PromoteArguments(CallGraphNode *CGN, CallGraph &CG,
     // not apply to inalloca.
     bool isSafeToPromote = (PtrArg->hasByValAttr() &&
        (isDenselyPacked(AgTy, DL) || !canPaddingBeAccessed(PtrArg)));
+
+    for (auto &U : F->uses()) {
+      CallSite CS{U.getUser()};
+      if (auto I = CS.getInstruction()) {
+        auto parent = I->getParent()->getParent();
+        if (sycl::isKernel(*parent)) {
+          isSafeToPromote = true;
+          DEBUG(dbgs() << "SYCL: " << parent->getName() << "is called in kernel.\n");
+          break;
+        }
+
+        // If the function has ancestor kernel, force to do argument promotion
+        std::vector<CallGraphNode *> Deps;
+        std::vector<CallGraphNode *> Done;
+        Deps.push_back(CG[parent]);
+
+        while (!Deps.empty()) {
+          CallGraphNode *node = Deps.back();
+          Deps.pop_back();
+          Done.push_back(node);
+          for (auto itr = node->begin(); itr != node->end(); itr++) {
+            if (auto ef = CallSite(itr->first).getCalledFunction()) {
+              if (sycl::isKernel(*ef)) {
+                DEBUG(dbgs() << "SYCL: " << ef->getName() << "has ancestor kernel.\n");
+                isSafeToPromote = true;
+                break;
+              }
+
+              if (std::find(Deps.begin(), Deps.end(),CG[ef]) != Deps.end())
+                Deps.push_back(CG[ef]);
+            }
+          }
+        }
+      }
+    }
+
```

Comment:
I am curious about the complexity of this method based on callee-graph compared to a more direct approach with a single DFS on the call graph starting from kernels and just adding the function to a set of called-from-kernel functions that you can just use later when required. This could go to lib/SYCL/SYCLKernel.cpp

Reply:
Try to move the code in DFS algorithm and mark all functions that have kernels as ancestors. And in SYCL argument flattening pass, we can easily find these marked function, and force their arguments flattened.